src/Client/QueryFuzzer.cpp

```diff
@@ -1351,14 +1356,59 @@ void QueryFuzzer::collectFuzzInfoRecurse(ASTPtr ast)
     }
 }
 
+template <typename T>
+std::optional<std::pair<T, T>> tryCast(ASTPtr & left, const ASTPtr & right)
+{
+    auto * left_cast = typeid_cast<T>(left.get());
+    auto * right_cast = typeid_cast<T>(right.get());
+
+    if (!left_cast || !right_cast)
+        return {};
+
+    return std::pair<T, T>(left_cast, right_cast);
+}
+
+void QueryFuzzer::crossover(ASTPtr main, const ASTPtr & donor)
+{
+    if (!main || !donor)
+        return;
+
+    if (auto res = tryCast<ASTSelectWithUnionQuery *>(main, donor); res.has_value())
+    {
+        auto [main_p, donor_p] = res.value();
+        crossover(main_p->list_of_selects, donor_p->list_of_selects);
+    }
+    else if (auto res_select = tryCast<ASTSelectQuery *>(main, donor); res_select.has_value())
+    {
+        auto [main_p, donor_p] = res_select.value();
+
+        for (auto expression : ASTSelectQuery::All)
+        {
+            if (!rollTheDice(2))
+                continue;
+
+            if (auto organ = donor_p->getExpression(expression, true))
+                main_p->setExpression(expression, std::move(organ));
+        }
+    }
+
+    for (auto & main_child : main->children)
+        for (const auto & donor_child : donor->children)
+            crossover(main_child, donor_child);
+
+}
+
 void QueryFuzzer::fuzzMain(ASTPtr & ast)
 {
     current_ast_depth = 0;
     debug_visited_nodes.clear();
     debug_top_ast = &ast;
 
     collectFuzzInfoMain(ast);
-    fuzz(ast);
+    if (coinToss())
```

Comment:
I'd probably give less probability of crossover since I'd expect it to be less useful (produce more broken queries) but it isn't a strong opinion.